core/patina_stacktrace/src/aarch64/unwind.rs

```diff
@@ -390,440 +404,449 @@ impl fmt::Display for UnwindCode {
-                // SaveR19R20X(u8),    // 001zzzzz
+                // SaveR19R20X(u8) -> 001zzzzz
+                // 001zzzzz: save <x19,x20> pair at [sp-#Z*8]!, pre-indexed offset >= -248
+
                 let z = byte & 0b00011111;
-                // log::info!("{}", UnwindCode::SaveR19R20X(z));
-                offset += z as u32 * 8u32;
-                len += 1;
+                log::debug!("    > {}", UnwindCode::SaveR19R20X(z)); // debug
+
+                prev_sp += z as u64 * 8; // pre increment the offset
+                // ignore r19 r20 values
+
+                i += 1;
             } else if (byte >> 6) & 0b11 == 0b01 {
-                // SaveFpLr(u8),       // 01zzzzzz |  stp   fp,lr,[sp,#0x{:X}]
+                // SaveFpLr(u8) -> 01zzzzzz |  stp   fp,lr,[sp,#0x{:X}]
+                // 01zzzzzz: save <x29,lr> pair at [sp+#Z*8], offset <= 504.
+
                 let z = byte & 0b00111111;
-                // log::info!("{}", UnwindCode::SaveFpLr(z));
-                // From the current offset move z bytes to get to <fp,lr>
-                lr_offset = offset + z as u32 * 8u32 + 8 /* step over fp */;
-                len += 1;
+                log::debug!("    > {}", UnwindCode::SaveFpLr(z)); // debug
+
+                // no pre decrement of sp
+                prev_fp = read_pointer64(prev_sp + z as u64 * 8); // dereference fp
+                prev_pc = read_pointer64(prev_sp + z as u64 * 8 + 8 /* step over fp */); // dereference lr
+
+                i += 1;
             } else if (byte >> 6) & 0b11 == 0b10 {
-                // SaveFpLrX(u8),      // 10zzzzzz | stp   fp,lr,[sp,#-0x{:X}]!
+                // SaveFpLrX(u8) -> 10zzzzzz | stp   fp,lr,[sp,#-0x{:X}]!
+                // 10zzzzzz: save <x29,lr> pair at [sp-(#Z+1)*8]!, pre-indexed offset >= -512
+
                 let z = byte & 0b00111111;
-                // log::info!("{}", UnwindCode::SaveFpLrX(z));
-                offset += (z as u32 + 1u32) * 8u32; // pre increment the offset
-                lr_offset = offset + 8 /* step over fp */;
-                len += 1;
+                log::debug!("    > {}", UnwindCode::SaveFpLrX(z)); // debug
+
+                prev_fp = read_pointer64(prev_sp); // dereference fp
+                prev_pc = read_pointer64(prev_sp + 8 /* step over fp */); // dereference lr
+                prev_sp += (z as u64 + 1) * 8; // pre increment the offset
+
+                i += 1;
             } else if (byte >> 3) & 0b11111 == 0b11000 {
-                // AllocM(u16),        // 11000xxx'xxxxxxxx
-                let x = (((byte & 0b111) as u16) << 8) | bytes[len + 1] as u16;
-                // log::info!("{}", UnwindCode::AllocM(x));
-                offset += x as u32 * 16u32;
-                len += 2;
+                // AllocM(u16) -> 11000xxx'xxxxxxxx
+                // 11000xxx'xxxxxxxx: allocate large stack with size < 32K (2^11 * 16).
+
+                let x = (((byte & 0b111) as u16) << 8) | unwind_codes[i + 1] as u16;
```

Comment:
Have you checked the bounds of all the code inside this while loop?

For example, the loop condition is `while i < unwind_codes.len()`. So, if `unwind_code.len()` is `5`, `i` is `4`, `i + 1` would access `unwind_codes[5]`.

I'm not looking at the logic closely, just wanted to call this out and double check if you ensured it all makes sense.